### PR TITLE
Compute diagonal with MatrixFree

### PR DESCRIPTION
**Pull request by peterrum**
This PR adds a new function to compute the diagonal of a system matrix in matrix-free manner for systems with constraints.

I still have to move the function `MF::compute_diagonal` from the test to the right place and generalize it for vector quantities. Also, I am not that happy with the interface yet...

Ping @kronbichler @lpsaavedra @tjhei @tcclevenger

### CODE REVIEW

**Comment by peterrum** [comment]
@kronbichler The code now works also for vector quantities. The test @tcclevenger gave me in summer also passes (see tests/matrix_free/compute_diagonal_02.cc). I still need to get block vectors working (any suggestions for a setup so that I reach this line https://github.com/dealii/dealii/blob/master/include/deal.II/matrix_free/fe_evaluation.h#L4071 ?) and to extend the tests for parallel runs.

**Comment by peterrum** [comment]
Ready to be reviewed. We still need to decide on an appropriate place for the new method and an appropriate interface. As it is now it is somewhat clumpy... Maybe it might be the right time to introduce a *kernel-container*, which stores a reference to an instance of `MatrixFree` and `FEEvaluation` and has for CG a `cell_integral` method.

**Comment by peterrum** [comment]
@kronbichler: @lpsaavedra has copied the code into her test multigrid code and it seems to work as expected!

**Comment by peterrum** [comment]
>The algorithm looks correct to me. The question is how we want to proceed. I guess there would be value to integrate the content of the compute_diagonal function into the library, e.g. by a free function that takes a local_vmult() operation or similar. 

I would propose to introduce a matrix-free utility namespace. We could collect there `MatixFree`-related functions. The diagonal function could be the first function in that namespace, but I think there are many functions developed in our group, which would fit into that namespace (e.g. a utility function to setup the data structures for vector-version of `MatrixFree::reinit`).

Maybe, we could agree on a namespace name and a file name so that I can move the new function there and the review process could be started. 

Regarding the interface, I am open. Maybe other users of `MatrixFree` users could give some input!

> Another question would be if it makes sense to put some of the stuff you compute inside that function to the MatrixFree::reinit() function - but that of course depends on how expensive these operations really are and how often we anticipate that this function will be called. If the typical user calls it once, it makes little sense.

This strongly depends on the use case. If one wants to recompute the diagonal for the same constraint matrix, it might be useful to *cache* the information.

Anyway, I will make some performance tests in the next day, comparing the *new* way to compute the matrix diagonal and the naive way (which does not work for all constraint types) and then we will see how expensive it is to compute the transpose of the local constraint matrix on the the fly.

**Comment by peterrum** [comment]
Lets postpone this PR after the release. First experiments showed that the code is about 3 times slower than the simple implementation. I would like to improve this!

File: include/deal.II/matrix_free/tools.h [hunk 1]
```diff
@@ -27,8 +30,369 @@ DEAL_II_NAMESPACE_OPEN
...
+  compute_diagonal(
+    const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+    FEEvaluation<dim,
+                 fe_degree,
+                 n_q_points_1d,
+                 n_components,
+                 Number,
+                 VectorizedArrayType> &                 phi,
+    LinearAlgebra::distributed::Vector<Number> &        diagonal_global,
+    const std::function<void()> &                       local_vmult)
```
Comment (tcclevenger) [inline include/deal.II/matrix_free/tools.h:62]: Can we add input argument of an unsigned int here and pass the cell index below? I need to query a coefficient table in my `local_vmult()`. 
```suggestion
    const std::function<void(const unsigned int)>&                       local_vmult)
```

File: include/deal.II/matrix_free/tools.h [hunk 2]
```diff
@@ -27,8 +30,369 @@ DEAL_II_NAMESPACE_OPEN
...
+            for (unsigned int i = 0; i < phi.dofs_per_cell; ++i)
+              {
+                // compute i-th column of element stiffness matrix:
+                // this could be simply performed as done at the moment with
+                // matrix-free operator evaluation applied to a ith-basis vector
+
+                for (unsigned int j = 0; j < phi.dofs_per_cell; ++j)
+                  phi.begin_dof_values()[j] = static_cast<Number>(i == j);
+
+                local_vmult();
```
Comment (tcclevenger) [inline include/deal.II/matrix_free/tools.h:339]:
```suggestion
                local_vmult(cell);
```

File: tests/matrix_free/compute_diagonal_util.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                   n_components,
+                   Number,
+                   VectorizedArrayType>
+        phi(matrix_free);
+
+      VectorType diagonal_global;
+      MatrixFreeTools::compute_diagonal(matrix_free,
+                                        phi,
+                                        diagonal_global,
+                                        [&]() { this->cell_operation(phi); });
```
Comment (tcclevenger) [inline tests/matrix_free/compute_diagonal_util.h:86]:
```suggestion
                                        [&](const unsigned int /*cell*/) { this->cell_operation(phi); });
```

File: include/deal.II/matrix_free/tools.h (hunk 1 above)
Comment (peterrum) [reply include/deal.II/matrix_free/tools.h:176]: @tcclevenger Sorry for the late reply! Thanks for testing the code in ASPECT. I don't think we need the additional argument. A new function has been introduced recently, s.t. it is possible to get the index from `FEEvaluation`:

https://github.com/dealii/dealii/blob/040c6ad7d46f1291ab5c038b70cf1da4fb22b006/include/deal.II/matrix_free/fe_evaluation.h#L237-L241

That should also work, shouldn't it?

**Comment by peterrum** [comment]
I have extended the documentation. Furthermore, I have modified the interface of the function since it was not thread safe. Now a new FEEvaluation is created within the matrix-free loop.

File: include/deal.II/matrix_free/tools.h [hunk 4]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
   }
-} // namespace MatrixFreeTools
 
+  namespace internal
+  {
+    template <typename Number>
+    struct LocalCSR
+    {
+      std::vector<unsigned int> row_lid_to_gid;
+      std::vector<unsigned int> row{0};
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:156]: Can you remove the initialization from the struct declaration and initialize at first use instead? Alternatively use a constructor.

File: include/deal.II/matrix_free/tools.h [hunk 5]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
+
+    // initialize vector
+    matrix_free.initialize_dof_vector(diagonal_global);
+
+    int dummy;
+
+    matrix_free.template cell_loop<VectorType, int>(
+      [&](const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+          LinearAlgebra::distributed::Vector<Number> &        diagonal_global,
+          const LinearAlgebra::distributed::Vector<Number> &,
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:192]: shouldn't this argument be
```suggestion
          const int &,
```
to reflect the template argument?

File: include/deal.II/matrix_free/tools.h [hunk 6]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
+        // local storage: buffer so that we access the global vector once
+        // note: may be larger then dofs_per_cell in the presence of
+        // constraints!
+        std::array<std::vector<Number>, n_lanes> diagonals_local_constrained;
+
+        for (unsigned int cell = range.first; cell < range.second; ++cell)
+          {
+            /*****************************************************************
+             * STEP 1: get relevant information from FEEvaluation
+             ****************************************************************/
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:213]: Not that I feel strongly about it, but I think we typically do not use this heavy-weight separation within a single function in many other places. I think a simple
```suggestion
            // step 1: get relevant information from FEEvaluation
```
would suffice (similarly below).

File: include/deal.II/matrix_free/tools.h [hunk 7]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
+                next_index_indicators =
+                  dof_info
+                    .row_starts[(cell * n_lanes + v) * n_fe_components +
+                                first_selected_component + 1]
+                    .second;
+
+                /***************************************************************
+                 * STEP 2a: setup locally-relevant constraint matrix in a
+                 *   coordinate list (COO - see:
+                 *   https://en.wikipedia.org/wiki/Sparse_matrix#Coordinate_list_(COO))
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:260]: I do not think we need a link to COO here. One either understands coordinate list/COO or does not mess around in this code I believe.

File: include/deal.II/matrix_free/tools.h [hunk 8]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
+                                           c_pool.col.begin() +
+                                             c_pool.row[j + 1],
+                                           i);
+
+                        // explanation: j-th row of C_e^T is empty (see above)
+                        if (scale_iterator ==
+                            c_pool.col.begin() + c_pool.row[j + 1])
+                          continue;
+
+                        // explanation: C_e^T(j,i) is is zero (see above)
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:488]:
```suggestion
                        // explanation: C_e^T(j,i) is zero (see above)
```

File: include/deal.II/matrix_free/tools.h [hunk 9]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
+
+            // loop over all columns of element stiffness matrix
+            for (unsigned int i = 0; i < phi.dofs_per_cell; ++i)
+              {
+                // compute i-th column of element stiffness matrix:
+                // this could be simply performed as done at the moment with
+                // matrix-free operator evaluation applied to a ith-basis vector
+
+                for (unsigned int j = 0; j < phi.dofs_per_cell; ++j)
+                  phi.begin_dof_values()[j] = static_cast<Number>(i == j);
```
Comment (kronbichler) [inline include/deal.II/matrix_free/tools.h:461]: I would be curious how the compiler translates this code. :wink:

**Review by kronbichler — APPROVED** [review]
Looks good in general.

**Comment by kronbichler** [comment]
Regarding the old question whether to integrate the transposition in the matrix-free setup and
> Anyway, I will make some performance tests in the next day, comparing the new way to compute the matrix diagonal and the naive way (which does not work for all constraint types) and then we will see how expensive it is to compute the transpose of the local constraint matrix on the the fly.

I think we can postpone that part to see how much it affects run time.

File: include/deal.II/matrix_free/tools.h [hunk 10]
```diff
@@ -119,8 +146,381 @@ namespace MatrixFreeTools
...
+              for (unsigned int v = 0; v < n_lanes_filled; ++v)
+                dof_indices[v] =
+                  dof_info.dof_indices.data() +
+                  dof_info
+                    .row_starts[(cell * n_lanes + v) * n_fe_components +
+                                first_selected_component]
+                    .first;
+
+              for (unsigned int v = n_lanes_filled; v < n_lanes; ++v)
+                dof_indices[v] = nullptr;
```
Comment (tjhei) [inline include/deal.II/matrix_free/tools.h:233]: you can just initialize ``dof_indices`` to 0 above and skip this.

**Comment by peterrum** [comment]
@kronbichler @tjhei Thanks for your feedback! I have addressed the comments.

**Review by kronbichler — APPROVED** [review]
Thanks for the adjustments.

**Comment by peterrum** [comment]
Thanks! As a follow-up PR I will have a look at block systems.

**Pull request merged**